include/picongpu/plugins/saxs/Saxs.kernel

```diff
@@ -0,0 +1,291 @@
+        unsigned int n_qy,
+        unsigned int n_qz,
+        unsigned int n_q
+    ) const
+    {
+        using namespace mappings::threads;
+        namespace po = boost::program_options;
+
+        constexpr uint32_t frameSize =
+            pmacc::math::CT::volume<SuperCellSize>::type::value;
+        constexpr uint32_t numWorker = T_numWorkers;
+
+        using FrameType = typename ParBox::FrameType;
+        using FramePtr = typename ParBox::FramePtr;
+
+        uint32_t const workerIdx = threadIdx.x;
+
+        constexpr int blockSize =
+            pmacc::math::CT::volume<SuperCellSize>::type::value;
+
+        // vectorial part of the integrand in the Jackson formula
+        PMACC_SMEM(acc, r_s, memory::Array<picongpu::float3_64, blockSize>);
+
+        // storage for macro particle weighting
+        PMACC_SMEM(acc, saxsWeighting_s, memory::Array<float_X, blockSize>);
+
+        PMACC_SMEM(acc, np_s, float_64);
+
+        // get extent of guarding super cells (needed to ignore them)
+        DataSpace<simDim> const guardingSuperCells =
+            mapper.getGuardingSuperCells();
+
+        /* number of super cells on GPU per dimension (still including guard
+         * cells) remove both guards from count [later one sided guard needs to
+         * be added again]
+         */
+        DataSpace<simDim> const superCellsCount(
+            mapper.getGridSuperCells() - 2 * guardingSuperCells);
+
+        // get absolute number of relevant super cells
+        int const numSuperCells = superCellsCount.productOfComponents();
+
+        /* go over all super cells on GPU
+         * but ignore all guarding supercells
+         */
+        for (int super_cell_index = 0; super_cell_index <= numSuperCells;
+             ++super_cell_index)
+        {
+            // select SuperCell and add one sided guard again
+            DataSpace<simDim> const superCell =
+                DataSpaceOperations<simDim>::map(
+                    superCellsCount, super_cell_index) +
+                guardingSuperCells;
+
+            // guardingSuperCells remove guarding block
+            DataSpace<simDim> const superCellOffset(
+                globalOffset +
+                ((superCell - guardingSuperCells) * SuperCellSize::toRT()));
+
+            // pointer to frame storing particles
+            FramePtr frame = pb.getLastFrame(superCell);
+
+            // number of particles in current frame
+            lcellId_t particlesInFrame =
+                pb.getSuperCell(superCell).getSizeLastFrame();
+
+            // go to next supercell
+            while (frame.isValid())
+            {
+                /* since a race condition can occur if "continue loop" is
+                 * called, all threads must wait for the selection of a new
+                 * frame until all threads have evaluated "isValid"
+                 */
+                __syncthreads();
+
+                ForEachIdx<IdxConfig<1, numWorker>> onlyMaster{workerIdx};
```

Comment:
Suggested change of parallelism (4): replace `workerIdx` with `threadIdx.x` (or make a new variable for it and use it here).